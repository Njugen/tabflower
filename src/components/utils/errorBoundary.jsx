import React, {Component} from "react";
import ErrorOverlay from '../modals/errorOverlay';
import { PropTypes } from 'prop-types';

/*
  Catch errors in child component's lifecycle hooks, during rendering and in 
  component constructors. For each error caught, show a fallback UI.

  Read more about error boundaries in React:
  https://reactjs.org/docs/error-boundaries.html
*/

class ErrorBoundary extends Component {
    state = { errors: [] };

     /* Update state so the next render will show the fallback UI. */
    static getDerivedStateFromError(error) {

      let currentErrors = [];
      const newError = error;
      currentErrors.push(newError);

      return { errors: currentErrors };
    }
    
    /*
      Catch an error and add it as an object in the this.state.error array for later use.
    */
    componentDidCatch(error, errorInfo) {
        let currentErrors = this.state.errors;
        const newError = Object.assign(error, errorInfo);
        currentErrors.push(newError);
        
        this.setState({
            errors: currentErrors
        }, () => {
            
        })
    }
  
    render() {
      if (this.state.errors.length > 0) {
        /*
          If errors are stored in the boundary state, render the ErrorOverlay UI component and inform the user about the errors
          using the information stored in this.state.errors
        */
        return <ErrorOverlay data={this.state.errors} onSave={() => ""} onDismiss={() => window.location.reload()}></ErrorOverlay>
      }
  
      return this.props.children || null; 
    }
  }

  ErrorBoundary.propTypes = {
    children: PropTypes.any.isRequired
  }

  export default ErrorBoundary;