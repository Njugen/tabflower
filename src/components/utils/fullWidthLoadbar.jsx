import React, { Component } from "react";

class FullWidthLoadbar extends Component {
    origin = {
        fillerWidth: 0,
        loadbarIsFilled: false,
        transitionCompleted: false,
        refreshFactor: this.props.refreshFactor
    }

    state = {
        ...this.origin
    }

    componentDidMount = () => {
        this.increaseFillerWidth();
    }
    
    componentDidUpdate = (prevProps, prevState) => {
        console.log("");
        if(this.props.refreshFactor !== prevState.refreshFactor){
            this.origin.refreshFactor = this.props.refreshFactor;

            this.increaseFillerWidth();
        }
    }

    resetState = (callback) => {
        const newState = {
            ...this.origin
        };
        
        this.setState(newState, callback);
    }

    increaseFillerWidth = () => {
        // Fill the loadbar by resetting the component's state, and then loop a percentage value using an interval
        let fillerWidth = 0;
       
        this.resetState(
            () => {
                const interval = setInterval(
                    () => {
                         if(fillerWidth < 100){
                             fillerWidth++;

                             this.setState({
                                 fillerWidth: fillerWidth
                             });
                         } else {
                             this.setState({
                                 fillerWidth: 0,
                                 loadbarIsFilled: true
                             });

                             clearInterval(interval);
                         }
                    },
                    1
                );
            }
        );

        
    }

    

    render = () => {
        const { 
            fillerWidth, 
            loadbarIsFilled,
            transitionCompleted
        } = this.state;

        // Inline style for the loadbar container itself
        const loadbarStyle = {
            display: transitionCompleted === true || loadbarIsFilled === true ? "none" : "block"
        }

        // Inline style for the filler
        const fillerStyle = {
            width: fillerWidth + "%"
        }

        return (
            <div id="tabeon-loadbar" style={loadbarStyle}>
                <div id="filler" style={fillerStyle}>
                   
                </div>
            </div>
        );
    }

}

export default FullWidthLoadbar;