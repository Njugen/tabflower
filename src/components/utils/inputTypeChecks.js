const isEmptyString = (input) => {
    if(typeof input === "string"){
        if(input.length === 0 || input.replace(/ /g, "").length === 0){
            return true;
        } else {
            false
        }
    } else {
        return false;
    }
}