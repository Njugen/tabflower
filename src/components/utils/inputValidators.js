/*
    isObject()

    Check if the input is an object, 
    input = {}

    Params:
    - input (can be whatever)

    Returns:
    - true, if the input is an object
    - false, if the input is not an object
*/

const isObject = (input) => {
    if(typeof input === "object"){
        return true;
    } else {
        return false;
    }
} 

/*
    isArray

    Check if the input is an array.

    Params:
    - input (can be whatever)

    Returns:
    - true, if input is an array
    - false, if input is not an array or if input is not defined
*/

const isArray = (input) => {
    if(typeof input !== "undefined"){
        return Array.isArray(input);
    } else {
        return false;
    }
}

/*
    isEmptyString()

    Check if a value is an empty string. A value is an empty
    string if all of the fllowing conditions are met:
        - The input is a string
        - The input's length is 0 OR
        - The input only consists of empty space

    Params:
    - input (can be whatever)

    Returns: 
        - true, if the input is an empty string
        - false, if the input is not an empty string
*/

const isEmptyString = (input) => {
    if(typeof input === "string"){
        if(input.length === 0 || input.replace(/ /g, "").length === 0){
            return true;
        } else {
            return false
        }
    } else {
        return false;
    }
}

/*
    isString()

    Check if the input value is a string, which consists of more than 0 characters (not empty).
    
    Params:
    - input (can be whatever)
    
    Returns:
    - true, if input is string and longer than 0 characters
    - false, if the input is not a string OR if the input is a string consisting of 0 characters
*/

const isString = (input) => {
    if(typeof input === "string"){
        if(input.length > 0){
            return true;
        } else {
            return false
        }
    } else {
        return false;
    }
}

/*
    isNumber()

    Verify that the input is actually a number

    Parameters:
    - input (Can be whatever)

    Returns:
    - true, if the input is a number (any number)
    - false, if the input is not a number
*/
const isNumber = (input) => {
    if(typeof input === "number"){
        return true;
    } else {
        return false
    }
}

/*
    isPositiveNumber()

    Check whether the input is a positive number or not

    Params:
    - input (can be anything)

    Returns:
    - true, if the input is a number AND the input is larger than 0
    - false, if the input is a number but not larger than 0
    - false, if the input is not a number
*/
const isPositiveNumber = (input) => {
    if(typeof input === "number"){
        if(input > 0){
            return true;
        } else {
            return false;
        }
    } else {
        return false;
    }
}

/*
    isNegativeNumber()

    Check if the input is a negative number

    Params:
    - input (can be any value)

    Returns:
    - true, if the input is a number and less than 0
    - false, if the input is a number, but larger than 0
    - false, if the input is not a number
*/
const isNegativeNumber = (input) => {
    if(typeof input === "number"){
        if(input < 0){
            return true;
        } else {
            return false;
        }
    } else {
        return false;
    }
}

