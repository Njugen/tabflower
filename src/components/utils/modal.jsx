import React, { Component, createRef } from "react";

class Modal extends Component {
    state = {
        data: {"hej": "ohja"},
        visibility: "hidden"
    }

    setVisibility = (state) => {
        const visibility = state;

        this.setState({
            ...this.state, visibility
        })
    }

    

    fadeIn = () => {
        let modal = this.modalRef.current;

        modal.style.opacity = 1;
        modal.style.zIndex = 10000;

      
    }

    fadeOut = () => {
        let modal = this.modalRef.current;

        modal.style.opacity = 0;
        setTimeout(() => {
            modal.style.zIndex = 0;
        }, 500)
        
        
        this.setVisibility("hidden");
    }

    cleanModalData = (callback) => {
        /*
            This function removes all data from the data section of
            the modal state. 

            This function also resets the modal UI, to make it ready for next use. Run this function
            when dismissing/saving the modal
        */
        this.setState({ data: {} }, () => {
            this.setVisibility("hidden");
            callback();
        });
    }

    dismissModalHandler = () => {
        this.cleanModalData(() => {
            this.fadeOut();
        })
        
    }

    saveDataHandler = () => {
        /*
            Save data function. We have three options:
            - Save data to modal by setting state (not recommended. Modal is washed cleaned at dismissal)
            - Save data to module by using props (not recommended. Module and state are reset at view switch)
            - Save data to view by using multilevel props (not recommended, code gets spaghettified, and view state are resetted at view switch)
            - Save data directly to backend or document by using service API/bridge
        */

        this.cleanModalData(() => {
            this.fadeOut();
        })
    }

    componentDidMount = () => {
        
        setTimeout(() => {
            this.fadeIn();
        }, 100)
    };

    componentDidUpdate = (prevProps, prevState) => {
        if(prevProps !== this.props){
            setTimeout(() => {
                this.fadeIn();
            }, 100)
        }
    }

    
    render = () => {
        
        return(
            <div ref={this.modalRef} className="modal fade" id="tabeonModal" tabIndex="-1" role="dialog" aria-labelledby="tabeonModalLabel" aria-hidden="true">
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title" id="tabeonModalLabel">Modal title</h5>
                            <button type="button" className="close" data-dismiss="modal" aria-label="Close" onClick={() => this.dismissModalHandler()}>
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            ...
                        </div>
                        <div className="modal-footer">
                            <button type="button" className="btn btn-secondary" data-dismiss="modal" onClick={() => this.dismissModalHandler()}>Close</button>
                            <button type="button" className="btn btn-primary" onClick={() => this.saveDataHandler()}>Save changes</button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }

    constructor(props){
        super(props);
        this.modalRef = createRef();
    }
    
}

export default Modal;