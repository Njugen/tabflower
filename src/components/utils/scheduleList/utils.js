const getWeekdaysArray = () => {
  let weekdays = [];

  weekdays[0] = "sunday";
  weekdays[1] = "monday";
  weekdays[2] = "tuesday";
  weekdays[3] = "wednesday";
  weekdays[4] = "thursday";
  weekdays[5] = "friday";
  weekdays[6] = "saturday";

  return weekdays;
};

export const getTodayObj = () => {
  const date = new Date();
  const weekdays = getWeekdaysArray();

  const day = date.getDay();

  return {
    dayName: weekdays[day],
    order: day,
  };
};

export const getUpcomingDayObj = (dayName) => {
  const weekdays = getWeekdaysArray();

  return {
    dayName,
    order: weekdays.indexOf(dayName),
  };
};

export const determineDayDifference = (today, futureDay) => {
  if (today > 6) return false;

  let n = 0;

  if (today >= futureDay) {
    n += 7;
  } else {
    n = 0;
  }

  const diff = today !== futureDay ? (futureDay + n - today) % 7 : 7;

  return diff;
};

export const determineFutureDate = (numOfDaysFromNow) => {
  const currentDateAsMS = Date.now();
  const numOfDaysAsMS = 60 * 60 * 24 * 1000 * numOfDaysFromNow;

  const futureDate = new Date(currentDateAsMS + numOfDaysAsMS);

  const startOfFutureDate = new Date(
    futureDate.getFullYear(),
    futureDate.getMonth(),
    futureDate.getDate(),
    0,
    0,
    0
  );

  return {
    date: startOfFutureDate.getDate(),
    month: startOfFutureDate.getMonth(),
    year: startOfFutureDate.getFullYear(),
    startOfDateTimeInMS: startOfFutureDate.getTime(),
  };
};

export const convertTimeToMS = (hour, minute) => {
  const hourParamInMS = hour * 60 * 60 * 1000;
  const minuteParamInMS = minute * 60 * 1000;

  return hourParamInMS + minuteParamInMS;
};
