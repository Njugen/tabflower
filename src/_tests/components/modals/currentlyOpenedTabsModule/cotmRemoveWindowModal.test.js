import React, { Fragment } from 'react';
import { shallow, mount, render } from 'enzyme';
import COTMRemoveWindowModal from './../../../../components/modals/currentlyOpenedTabsModule/cotmRemoveWindowModal';


import * as ExceptionsHandler from '../../../../components/utils/exceptionsAndHandler';
import * as validator from '../../../../components/utils/inputValidators';


const predefinedComponent = (props, options) => {
    props = props || {};

    const component = shallow(<COTMRemoveWindowModal {...props} />, options);
    return component;
}

let presetProps = {
    data: { },
    onRaiseToErrorOverlay: "",
    onDismiss: ""
};

let testComponent;
let componentInstance;

describe("Test <COTMRemoveWindowModal /> component behaviour at mount", () => {
    

    const actualErrorReturns = {
        "COTMRemoveWindowModal-101": ExceptionsHandler.ValidatorError("COTMRemoveWindowModal-101"),
        "COTMRemoveWindowModal-102": ExceptionsHandler.ValidatorError("COTMRemoveWindowModal-102"),
        "COTMRemoveWindowModal-103": ExceptionsHandler.ValidatorError("COTMRemoveWindowModal-103"),
        "COTMRemoveWindowModal-104": ExceptionsHandler.ValidatorError("COTMRemoveWindowModal-104")
    };

    const expectedErrorReturns = {
        "COTMRemoveWindowModal-101": {
            name: "ValidatorError",
            message: "The targetted window could not be identified since its unique identification number is missing or provided in wrong format. This might be caused if the window information has been modified per request. Please, make sure the window information has not been modified after being retrieved directly from the browser. As a result, the targetted window cannot be deleted at this time.",
            code: "COTMRemoveWindowModal-101"
        },
        "COTMRemoveWindowModal-102": {
            name: "ValidatorError",
            message: "No specific information about the targetted window could be retrieved, therefore the window cannot be closed at this point.",
            code: "COTMRemoveWindowModal-102"
        },
        "COTMRemoveWindowModal-103": {
            name: "ValidatorError",
            message: "The callback function is missing.",
            code: "COTMRemoveWindowModal-103"
        },
        "COTMRemoveWindowModal-104": {
            name: "ValidatorError",
            message: "No information about the targetted window could be retrieved, therefore the window cannot be closed at this point.",
            code: "COTMRemoveWindowModal-104"
        }
    }

    beforeEach(() => {
        jest.clearAllMocks();
        jest.useRealTimers();

        const presetProps = {
          
        };
        testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
        componentInstance = testComponent.instance();

        ExceptionsHandler.ErrorHandler = jest.fn();
        ExceptionsHandler.ValidatorError = jest.fn();
        ExceptionsHandler.ValidatorError.mockImplementation(errCode => {
            return actualErrorReturns[errCode];
        });
    });

    describe("Test verifyChildProps()", () => {
        const various_nonObjects = [
            ["a very weird looking text string"],
            [77],
            [false],
            [true],
            [undefined],
            [[1,2,3,4]],
            [() => {}],
            [null]
        ];

        describe("Examine the data object of this.props", () => {
            test("Run verifyChildProps(): If this.props.data is missing, throw an error \"COTMRemoveWindowModal-104\"", () => {
                 expect(() => {
                    const presetProps = {
                        
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-104"]); 
            });

            test.each(various_nonObjects)("Run verifyChildProps(): If this.props.data = %p (is not an object), throw an error \"COTMRemoveWindowModal-104\"", (val) => {
                expect(() => {
                    const presetProps = {
                        data: val
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-104"]);
            });

            test("Run verifyChildProps(): If \"params\" is missing in this.props.data, throw an error \"COTMRemoveWindowModal-104\"", () => {
                expect(() => {
                    const presetProps = {
                        data: {}
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-104"]);
            });

            test.each(various_nonObjects)("Run verifyChildProps(): If \"params\" = %p (is not an object) in in this.props.data, throw an error \"COTMRemoveWindowModal-104\"", (val) => {
                expect(() => {
                    const presetProps = {
                        data: {
                            params: val
                        }
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-104"]);
            }); 
            
        }); 

        describe("Examine the \"windowInfo\" of this.props.data.params", () => {
            test.each(various_nonObjects)("Run verifyChildProps(): If \"windowInfo\" = %p (is not an object) in in this.props.data.params, throw an error \"COTMRemoveWindowModal-102\"", (val) => {
                const presetProps = {
                    data: {
                        params: {
                            windowInfo: val
                        }
                    }
                };

                expect(() => {
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-102"]);
            });

            test("Run verifyChildProps(): If \"windowInfo\" is missing in this.props.data.params, throw an error \"COTMRemoveWindowModal-102\"", () => {
                const presetProps = {
                    data: {
                        params: {
                      
                        }
                    }
                };

                expect(() => {
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();

                    componentInstance.verifyChildProps();
                }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-102"]);
            });

            describe("Examine this.props.data.params.windowInfo.id", () => {
                const various_nonNumber = [
                    ["a very weird looking text string"],
                    [{ item: 123 }],
                    [false],
                    [true],
                    [undefined],
                    [[1,2,3,4]],
                    [() => {}],
                    [null]
                ]

                test("Run verifyChildProps(): If \"id\" is not provided by this.props.data.params.windowInfo, throw an error \"COTMRemoveWindowModal-101\"", () => {
                    const presetProps = {
                        data: {
                            params: {
                                windowInfo: {}
                            }
                        }
                    };
    
                    expect(() => {
                        testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                        componentInstance = testComponent.instance();
    
                        componentInstance.verifyChildProps();
                    }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-101"]);
                });

                test.each(various_nonNumber)("Run verifyChildProps(): If the \"id\" = %p (provided by this.props.data.params.windowInfo is not a number), throw an error \"COTMRemoveWindowModal-101\"", (val) => {
                    const presetProps = {
                        data: {
                            params: {
                                windowInfo: {
                                    id: val
                                }
                            }
                        }
                    };
    
                    expect(() => {
                        testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                        componentInstance = testComponent.instance();
    
                        componentInstance.verifyChildProps();
                    }).toThrow(expectedErrorReturns["COTMRemoveWindowModal-101"]);
                });
            })
            
        }); 
        
    });

    describe("Test childComponentMount()", () => {
        const various_nonObjects = [
            ["a very weird looking text string"],
            [77],
            [false],
            [true],
            [undefined],
            [[1,2,3,4]],
            [() => {}],
            [null]
        ];

        describe("Examine the \"data\" object of this.props", () => {
            test("Run childComponentDidMount(): if this.props.data object is missing, throw an error \"COTMRemoveWindowModal-104\"", () => {
                const presetProps = {
               
                };
                testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                componentInstance = testComponent.instance();
                componentInstance.childComponentDidMount();
    
                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-104");
            });
    
            test.each(various_nonObjects)("Run childComponentDidMount(): if this.props.data = %p (is not an object), throw an error \"COTMRemoveWindowModal-104\"", (val) => {
                const presetProps = {
                    data: val
                };
                testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                componentInstance = testComponent.instance();
                componentInstance.childComponentDidMount();
    
                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-104");
            });

            describe("Examine the \"params\" object of this.props.data", () => {
                test("Run childComponentDidMount(): if this.props.data.params object is missing, throw an error \"COTMRemoveWindowModal-104\"", () => {
                    const presetProps = {
                        data: {}
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();
                    componentInstance.childComponentDidMount();
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-104");
                });
        
                test.each(various_nonObjects)("Run childComponentDidMount(): if this.props.data.params = %p (is not an object), throw an error \"COTMRemoveWindowModal-104\"", (val) => {
                    const presetProps = {
                        data: {
                            params: val
                        }
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();
                    componentInstance.childComponentDidMount();
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-104");
                });
            });

            describe("Examine the \"windowInfo\" object of this.props.data.params", () => {
                test("Run childComponentDidMount(): If this.props.data.params.windowInfo is missing, throw an error \"COTMRemoveWindowModal-102\"", () => {
                    const presetProps = {
                        data: {
                            params: {
                                
                            }
                        }
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();
                    componentInstance.childComponentDidMount();
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-102");
                });

                test.each(various_nonObjects)("Run childComponentDidMount(): If this.props.data.params.windowInfo = %p (is not an object), throw an error \"COTMRemoveWindowModal-102\"", (val) => {
                    const presetProps = {
                        data: {
                            params: {
                                windowInfo: val
                            }
                        }
                    };
                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();
                    componentInstance.childComponentDidMount();
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-102");
                });

                test.each(various_nonObjects)("Run childComponentDidMount(): If this.props.data.params.windowInfo is an object, pass it to this.setState()", (val) => {
                    const presetProps = {
                        data: {
                            params: {
                                windowInfo: {}
                            }
                        }
                    };

                    const data = presetProps.data.params.windowInfo;

                    testComponent = predefinedComponent(presetProps, { disableLifecycleMethods: true });
                    componentInstance = testComponent.instance();
                    componentInstance.setState = jest.fn();
                    componentInstance.childComponentDidMount();
        
                    expect(componentInstance.setState).toHaveBeenCalledWith({ data });
                });
            })
        })
        
    })

    describe("Test saveModalHandler(callback)", () => {
        const various_nonFunctions = [
            ["a very weird looking text string"],
            [77],
            [false],
            [true],
            [undefined],
            [[1,2,3,4]],
            [{ key: "value" }],
            [null]
        ];

        test("Run saveModalHandler(): Throw an error \"COTMRemoveWindowModal-103\"", () => {
            componentInstance.saveModalHandler();

            expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-103");
        });

        test.each(various_nonFunctions)("Run saveModalHandler(%p): Throw an error \"COTMRemoveWindowModal-103\"", (val) => {
            componentInstance.saveModalHandler(val);

            expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("COTMRemoveWindowModal-103");
        });

        test("Run saveModalHandler(callbackFunction), where callbackFunction is a function: The callbackFunction() should get called, with this.state as its payload", () => {
            const callbackFunction = jest.fn();
            componentInstance.saveModalHandler(callbackFunction);

            expect(callbackFunction).toHaveBeenCalledWith(componentInstance.state);
        });
    });

    describe("Test dismissModalHandler()", () => {
        test("Run dismissModalHandler(): the function this.clearModalData() should be called", () => {
            componentInstance.clearModalData = jest.fn();
            componentInstance.dismissModalHandler();

            expect(componentInstance.clearModalData).toHaveBeenCalledTimes(1);
        })
    });

    describe("Test renderModalBody()", () => {

        describe("Examine the data object of this.props", () => {
            test("Run renderModalBody(): Whatever it returns should match the snapshot", () => {

                expect(componentInstance.renderModalBody()).toMatchSnapshot();
            });
        }); 

    });

    describe("Test renderModalHeader()", () => {
        test("Run renderModalHeader(): The function should return \"Close Window\"", () => {
            expect(componentInstance.renderModalHeader()).toBe("Close Window");
        });
    });

}) 