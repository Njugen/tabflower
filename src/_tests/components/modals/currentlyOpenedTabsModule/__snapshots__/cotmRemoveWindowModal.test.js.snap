// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <COTMRemoveUnresponsiveTabs /> component behaviour at mount Test renderModalBody() Examine the data object of this.props Run renderModalBody(): Whatever it returns should match the snapshot 1`] = `
<React.Fragment>
  <p>
    Closing this window will also close all its tabs. All ongoing activities on these web pages will be interrupted and possibly lost.
  </p>
  <p>
    Are you sure you want to proceed?
  </p>
  <p
    className="small"
  >
    You may reopen the window with all its tabs intact from the browser's history feature (presuming you have it activated).
  </p>
</React.Fragment>
`;

exports[`Test <COTMRemoveWindowModal /> component behaviour at mount Test renderModalBody() Examine the data object of this.props Run renderModalBody(): Whatever it returns should match the snapshot 1`] = `
<React.Fragment>
  <p>
    Closing this window will also close all its tabs. All ongoing activities on these web pages will be interrupted and possibly lost.
  </p>
  <p>
    Are you sure you want to proceed?
  </p>
  <p
    className="small"
  >
    You may reopen the window with all its tabs intact from the browser's history feature (presuming you have it activated).
  </p>
</React.Fragment>
`;
